From 3164581d5eb5a91c33948d43e19e1a38dc555d00 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Tue, 22 Sep 2020 12:24:13 -0700
Subject: [PATCH 1637/1851] v4l: xilinx: vpss-csc: Fix warnings for
 NULL_RETURNS

commit 7883bdc89343b1a0a4c7dcf34aa7b101898c5694 from
https://github.com/Xilinx/linux-xlnx.git

This patch fix the warnings for the NULL_RETURNS type.

The actual warnings are:

For xcsc_get_format():
Event returned_null: 	"__xcsc_get_pad_format" returns "NULL".
Event dereference: 	Dereferencing "__xcsc_get_pad_format(xcsc, cfg,
fmt->pad, fmt->which)", which is known to be "NULL"

For xcsc_set_format()
Event returned_null: 	"__xcsc_get_pad_format" returns "NULL".
Event var_assigned: 	Assigning: "__format" = "NULL" return value from
"__xcsc_get_pad_format".
Event dereference: 	Dereferencing "__format", which is known to be
"NULL".

For __xcsc_get_pad_format():
Event return_null: 	Explicitly returning null.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../media/platform/xilinx/xilinx-vpss-csc.c   | 27 ++++++++++++++++---
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-vpss-csc.c b/drivers/media/platform/xilinx/xilinx-vpss-csc.c
index c628ee07a1e0..752e7b4f047b 100644
--- a/drivers/media/platform/xilinx/xilinx-vpss-csc.c
+++ b/drivers/media/platform/xilinx/xilinx-vpss-csc.c
@@ -542,14 +542,22 @@ __xcsc_get_pad_format(struct xcsc_dev *xcsc,
 		      struct v4l2_subdev_pad_config *cfg,
 		      unsigned int pad, u32 which)
 {
+	struct v4l2_mbus_framefmt *format;
+
 	switch (which) {
 	case V4L2_SUBDEV_FORMAT_TRY:
-		return v4l2_subdev_get_try_format(&xcsc->xvip.subdev, cfg, pad);
+		format = v4l2_subdev_get_try_format(&xcsc->xvip.subdev,
+						    cfg, pad);
+		break;
 	case V4L2_SUBDEV_FORMAT_ACTIVE:
-		return &xcsc->formats[pad];
+		format = &xcsc->formats[pad];
+		break;
 	default:
-		return NULL;
+		format = NULL;
+		break;
 	}
+
+	return format;
 }
 
 static void
@@ -760,8 +768,13 @@ static int xcsc_get_format(struct v4l2_subdev *subdev,
 			   struct v4l2_subdev_format *fmt)
 {
 	struct xcsc_dev *xcsc = to_csc(subdev);
+	struct v4l2_mbus_framefmt *format;
 
-	fmt->format = *__xcsc_get_pad_format(xcsc, cfg, fmt->pad, fmt->which);
+	format = __xcsc_get_pad_format(xcsc, cfg, fmt->pad, fmt->which);
+	if (!format)
+		return -EINVAL;
+
+	fmt->format = *format;
 	return 0;
 }
 
@@ -774,9 +787,15 @@ static int xcsc_set_format(struct v4l2_subdev *subdev,
 	struct v4l2_mbus_framefmt *__propagate;
 
 	__format = __xcsc_get_pad_format(xcsc, cfg, fmt->pad, fmt->which);
+	if (!__format)
+		return -EINVAL;
+
 	/* Propagate to Source Pad */
 	__propagate = __xcsc_get_pad_format(xcsc, cfg,
 					    XVIP_PAD_SOURCE, fmt->which);
+	if (!__propagate)
+		return -EINVAL;
+
 	*__format = fmt->format;
 
 	__format->width = clamp_t(unsigned int, fmt->format.width,
-- 
2.31.1

